### PR TITLE
[LSP] Completions and definitions

### PR DESCRIPTION
**Pull request by artagnon**
Here's an early preview of commits based on `artagnon_lsp-wireframe` (#1742) -- the wireframe needs to be reviewed and merged before too much work is based on top of it.

### CODE REVIEW

**Comment by artagnon** [comment]
There seems to be a bug in F\*'s extraction mechanism due to which b654096 fails: many runtime errors of the form `(error "line 328 column 14: invalid sort declaration, sort already declared/defined")` when attempting to check `Prims.fst`. I don't have a reduced case yet.

**Comment by artagnon** [comment]
I found the bug!

```
let initial_repl_state : repl_state = (* Some impure function *)
let start_server () : unit = initial_repl_state; exit 0
```

Essentially, the type annotation on `initial_repl_state` is incorrect: it must be `unit -> repl_state`, not `repl_state`. Can someone who knows a little more F\* file a bug report of a reduced case?

**Comment by catalin-hritcu** [comment]
@victor-dumitrescu might be able to help

**Comment by artagnon** [comment]
- [X] Lax-check entire file on open
- [X] Diagnostic reporting, with location information
- [X] Lax-check entire file on save
- [X] Go to definition
- [X] Hover information: type and documentation
- [X] Auto-complete

**Comment by nikswamy** [comment]
Hi Ram, Can you provide some instructions here on how to test this with vsfstar? Thanks, Nik

**Comment by nikswamy** [comment]
Can you say more about this remark above? I'm not able to find the context for the bug.


>let initial_repl_state : repl_state = (* Some impure function *)
let start_server () : unit = initial_repl_state; exit 0

>Essentially, the type annotation on initial_repl_state is incorrect: it must be unit -> repl_state, not repl_state. Can someone who knows a little more F* file a bug report of a reduced case?

**Comment by artagnon** [comment]
> Hi Ram, Can you provide some instructions here on how to test this with vsfstar? Thanks, Nik

Build this branch, and put `fstar.exe` in path. Clone [vsfstar](https://github.com/artagnon/vsfstar), run `npm install` in the cloned directory, and run it from within vscode (Debug > Start Debugging).

**Comment by artagnon** [comment]
> Can you say more about this remark above? I'm not able to find the context for the bug.

The bug is triggered when I have an impure function with the following (incorrect) type annotation:

```
let initial_repl_state : repl_state = (* Some impure function with incorrect type annotation *)
let start_server () : unit = initial_repl_state; exit 0
```

Instead of reporting the fact that I have an incorrect type annotation, I instead get many confusing errors of the form `(error "line 328 column 14: invalid sort declaration, sort already declared/defined")`. When I fix the type annotation to say `unit -> repl_state`, the errors go away.

**Comment by artagnon** [comment]
@nikswamy Hi Nik, any progress on this? Is there anything I can do to help?

**Comment by aseemr** [comment]
Hi @artagnon, can you please provide some more instructions on how can I test it.

I installed `vscode` and `vsfstar`, the latter with this output from cygwin:

```
aseemr@ASEEMR-SBOOK ~
$ git clone git@github.com:artagnon/vsfstar.git
Cloning into 'vsfstar'...
remote: Enumerating objects: 141, done.
remote: Counting objects: 100% (141/141), done.
remote: Compressing objects: 100% (91/91), done.
remote: Total 141 (delta 69), reused 106 (delta 38), pack-reused 0
Receiving objects: 100% (141/141), 46.86 KiB | 266.00 KiB/s, done.
Resolving deltas: 100% (69/69), done.

aseemr@ASEEMR-SBOOK ~
$ cd vsfstar/

aseemr@ASEEMR-SBOOK ~/vsfstar
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean

aseemr@ASEEMR-SBOOK ~/vsfstar
$ npm install

> vsfstar@0.0.1 postinstall C:\cygwin64\home\aseemr\vsfstar
> node ./node_modules/vscode/bin/install

Detected VS Code engine version: ^1.33.0
Found minimal version that qualifies engine range: 1.33.0
Fetching vscode.d.ts from: https://raw.githubusercontent.com/Microsoft/vscode/0dd516dd412d42323fc3464531b1c715d51c4c1a/src/vs/vscode.d.ts
vscode.d.ts successfully installed!

added 115 packages from 436 contributors and audited 185 packages in 8.888s
found 0 vulnerabilities


aseemr@ASEEMR-SBOOK ~/vsfstar
$
```

After this I opened `vscode` and added the root `FStar` folder to it. Then I opened `FStar.SMTEncoding.Encode.fs` and tried to follow the steps you outlined above (`Debug -> Start Debugging`), but it asks me to `Select Environment` as follows:

im1

Clicking on `More` shows me extensions to install, but searching for `vsfstar` does not show anything.

Sorry if this is a trivial question, I haven't used `vscode` before, and thanks for your patience with this PR!

**Comment by aseemr** [comment]
> > Can you say more about this remark above? I'm not able to find the context for the bug.
> 
> The bug is triggered when I have an impure function with the following (incorrect) type annotation:
> 
> ```
> let initial_repl_state : repl_state = (* Some impure function with incorrect type annotation *)
> let start_server () : unit = initial_repl_state; exit 0
> ```
> 
> Instead of reporting the fact that I have an incorrect type annotation, I instead get many confusing errors of the form `(error "line 328 column 14: invalid sort declaration, sort already declared/defined")`. When I fix the type annotation to say `unit -> repl_state`, the errors go away.

Can you also please point me to the instance of this issue, could not find it in `FStar.Interactive.Lsp.fs`.

**Comment by aseemr** [comment]
Should we also open a new PR for this, with base as `master` and to-be-merged branch as `artagnon_lsp-complete-def`, this one has base `artagnon_lsp-wireframe`, so it does not show me diffs w.r.t. `master`.

File: src/fstar/FStar.Interactive.Ide.fs [hunk 1]
```diff
@@ -98,11 +60,11 @@ let with_captured_errors' env sigint_handler f =
     Util.print_string "Interrupted"; None
 
   | Error (e, msg, r) ->
-    TcErr.add_errors env [(e, msg, r)];
+    Errors.add_errors [(e, msg, r)];
```
Comment (aseemr) [inline src/fstar/FStar.Interactive.Ide.fs]: Why is this change required?

File: src/fstar/FStar.Interactive.Ide.fs [hunk 2]
```diff
@@ -884,7 +615,7 @@ let collect_errors () =
   errors
 
 let run_segment (st: repl_state) (code: string) =
-  let frag = { frag_text = code; frag_line = 1; frag_col = 0 } in
+  let frag = { frag_fname = "<input>"; frag_text = code; frag_line = 1; frag_col = 0 } in
```
Comment (aseemr) [inline src/fstar/FStar.Interactive.Ide.fs]: Why are we passing `"<input>"` here, instead of `st.repl_fname`? Are these two difference?

File: src/fstar/FStar.JsonHelper.fs [hunk 3]
```diff
@@ -1,17 +1,23 @@
-(* Helpers mainly for FStar.Interactive.Lsp; some sharing with
- * FStar.Interactive.Ide *)
+(* Json helpers mainly for FStar.Interactive.Lsp; some sharing with *
```
Comment (aseemr) [inline src/fstar/FStar.JsonHelper.fs]: Can you please add the copyright notice to this and other new files? You can copy it from some other file that has it.

File: src/fstar/FStar.PushHelper.fs [hunk 4]
```diff
@@ -0,0 +1,367 @@
+(* FStar.Interactive.Lsp and FStar.Interactive.Ide need to push various *
```
Comment (aseemr) [inline src/fstar/FStar.PushHelper.fs]: Should its name be `FStar.Interactive.PushHelper.fs`, to say that it is used only by the interactive mode(s)? And how about `JsonHelper`, is that also interactive mode specific, or can it also be used by any other code that manipulates JSON?

File: src/fstar/FStar.Universal.fs [hunk 5]
```diff
@@ -139,9 +139,10 @@ let init_env deps : TcEnv.env =
 (* Interactive mode: checking a fragment of a code                     *)
 (***********************************************************************)
 let tc_one_fragment curmod (env:TcEnv.env_t) frag =
+  let file_from_env env = Range.file_of_range (TcEnv.get_range env) in
   let acceptable_mod_name modul =
     (* Interface is sent as the first chunk, so we must allow repeating the same module. *)
-    Parser.Dep.lowercase_module_name (List.hd (Options.file_list ())) =
+    Parser.Dep.lowercase_module_name (file_from_env env) =
```
Comment (aseemr) [inline src/fstar/FStar.Universal.fs]: Can you please explain this change too, why does the previous code not work for LSP? (A comment in the code would be useful.)

**Comment by aseemr** [comment]
Hi @artagnon, while we wait for the windows issue to be resolved, I have added some review comments. Many are requests for adding some comments to explain the changes to the non-LSP files (e.g. to `TypeChecker.TcTerm`). Can you please add some comments explaining those changes?

Also, documenting (in terms of comments just before each signature) at least the functions in the new `.fsi` files would be nice.

I am also guessing that you checked that this does not change the current emacs mode behavior? (There seems to some code churn around that.)

Thanks a lot, looking forward to trying it out!

**Comment by artagnon** [comment]
> > > Can you say more about this remark above? I'm not able to find the context for the bug.
> > 
> > 
> > The bug is triggered when I have an impure function with the following (incorrect) type annotation:
> > ```
> > let initial_repl_state : repl_state = (* Some impure function with incorrect type annotation *)
> > let start_server () : unit = initial_repl_state; exit 0
> > ```
> > 
> > 
> > Instead of reporting the fact that I have an incorrect type annotation, I instead get many confusing errors of the form `(error "line 328 column 14: invalid sort declaration, sort already declared/defined")`. When I fix the type annotation to say `unit -> repl_state`, the errors go away.
> 
> Can you also please point me to the instance of this issue, could not find it in `FStar.Interactive.Lsp.fs`.

See [this](https://github.com/FStarLang/FStar/pull/1744/files#diff-34f056c207c8944e0362623b80c56fc8R211). Please let me know if you need further clarifications.

**Comment by artagnon** [comment]
Current state: Works fine on macOS and Linux, but is a no-op on Windows (thanks to @aseemr @catalin-hritcu for testing this out). Some clarifications to the instructions are necessary:

1. Run [vsfstar](https://github.com/artagnon/vsfstar) from vscode using <kbd>F5</kbd>.
2. Install the "FStar Syntax highlighter" plugin from the vscode extension marketplace.
3. After building this branch of FStar, put the path to the binary inside settings.json (<kbd>Ctrl+Shift+P</kbd> > `Edit Settings (JSON)`) like this: ```
{
"fstar.path" : "C:/fstar/bin/fstar.exe"
}```
noting the use of forward slashes instead of backslashes.
4. There are a couple of `if false` branches in `FStar.Interactive.Lsp.fs` that can be changed to `if true` to enable debug printing. Note the output in the `View > Debug Console > Switch to Output tab > Select F* Language Server from the dropdown`.

I suspect I've baked the assumption of having only UNIX-like paths in the code, and am working on fixing this.

File: src/fstar/FStar.Interactive.Ide.fs (hunk 2 above)
Comment (artagnon) [reply src/fstar/FStar.Interactive.Ide.fs]: As it so happens, these are two different things. Ideally, there should be just one place where we keep track of the filename of the fragment, but this is not so in reality. In the interactive mode, there are different "modes of operation", including a mode where we send a "text fragment": in that case, the filename is `"<input>"`. In Lsp, there is a single mode of operation: we only send across fragments. To cope with this difference, we need this change.

File: src/fstar/FStar.JsonHelper.fs (hunk 3 above)
Comment (artagnon) [reply src/fstar/FStar.JsonHelper.fs]: Should I assign all copyrights to Nik?

File: src/fstar/FStar.PushHelper.fs (hunk 4 above)
Comment (artagnon) [reply src/fstar/FStar.PushHelper.fs]: JsonHelper can be used by any code that manipulates JSON, but PH and QH need the `Interactive` prefix. Thanks for catching this.

File: src/fstar/FStar.Universal.fs (hunk 5 above)
Comment (artagnon) [reply src/fstar/FStar.Universal.fs]: Lsp is started with `fstar.exe --lsp` without providing a single file in `Options.file_list ()`, while Interactive is started with `fstar.exe --ide filename.fst` providing a single file, making `hd` yield the desired result. I'll include a comment.

File: src/fstar/FStar.Interactive.Ide.fs (hunk 1 above)
Comment (artagnon) [reply src/fstar/FStar.Interactive.Ide.fs]: To be honest, I don't remember why I made that change. Whatever the original reason was, it no longer seems to be valid, since reverting the change doesn't break anything -- thanks a lot of catching this!

Comment (artagnon) [reply src/fstar/FStar.Interactive.Ide.fs]: I'm very sorry about this, but this change is required to make the Ide regression suite pass — I wish I could remember how the change ties in, but I can't; it's been almost 2 months since I posted this branch for review, and I can scarcely remember the details 😓

**Comment by artagnon** [comment]
> Hi @artagnon, while we wait for the windows issue to be resolved, I have added some review comments. Many are requests for adding some comments to explain the changes to the non-LSP files (e.g. to `TypeChecker.TcTerm`). Can you please add some comments explaining those changes?

Done.

> Also, documenting (in terms of comments just before each signature) at least the functions in the new `.fsi` files would be nice.

I've added some comments.

> I am also guessing that you checked that this does not change the current emacs mode behavior? (There seems to some code churn around that.)

The Ide regression suite seems to pass, fwiw. Could someone who's more familiar with the Ide mode + Emacs try this out, so we know that nothing is broken?

> Thanks a lot, looking forward to trying it out!

I still can't explain what's wrong with LSP on Windows. Moreover, I'm unclear on the next steps, since I don't have access to a Windows VM myself.

File: src/fstar/FStar.JsonHelper.fs (hunk 3 above)
Comment (nikswamy) [reply src/fstar/FStar.JsonHelper.fs]: Certainly not! : ) Some files have copyright notices from when I first wrote them, so those mention my name; most do not. I've updated the style guide with a sample copyright notice. https://github.com/FStarLang/FStar/wiki/Style-guide

Comment (aseemr) [reply src/fstar/FStar.JsonHelper.fs]: For example, you could copy it from here: https://github.com/FStarLang/FStar/blob/master/src/fstar/FStar.CheckedFiles.fsi.

**Comment by artagnon** [comment]
> > Thanks a lot, looking forward to trying it out!
> 
> I still can't explain what's wrong with LSP on Windows. Moreover, I'm unclear on the next steps, since I don't have access to a Windows VM myself.

Hi @aseemr, it now works on Windows 🎉 — please try it out; looking forward to a speedy merge.

File: INSTALL.md [hunk 6]
```diff
@@ -162,7 +162,13 @@ See [Running F\* from a docker image](https://github.com/FStarLang/FStar/wiki/Ru
...
 ## Building F\* from sources ##
 
-Short version: Simply run `make -C src -j6 ocaml-fstar-ocaml` from the `master` branch of the clone.
+Short version: Simply run `make -C src -j6 ocaml-fstar-ocaml` from the `master` branch of the clone. If you're developing a part of F\* that changes the OCaml output, you could get stuck due to a dirty worktree; in that case, run:
+
+```
+$ git clean -dfx
+$ git checkout -- src/ocaml-output
+$ make -C src -j6 ocaml-fstar-ocaml
+```
```
Comment (catalin-hritcu) [inline INSTALL.md]: How about moving this to `CONTRIBUTING.md`? It's not something a first time user of F* will have to worry about.

Also, how about adding a big warning explaining what `git clean -dfx` does. I sometimes do it for subdirs, never for the main dir.

**Comment by catalin-hritcu** [comment]
I tried this again and on Linux all is fine, but on Windows, while things have improved and Completion and "Peek Definition" work fine, "Go to Definition" still has problems. Screenshots below (on bottom left the errors I get when hitting F12):
Screenshot_2019-07-26_13-34-32
Screenshot_2019-07-26_13-38-49

**Comment by artagnon** [comment]
Thanks for testing it @catalin-hritcu! Unfortunately, I had to return the Windows box this morning. The goto-definition issue on Windows seems to be related to path encoding (again) — what do you propose we do?

**Comment by artagnon** [comment]
> Thanks for testing it @catalin-hritcu! Unfortunately, I had to return the Windows box this morning. The goto-definition issue on Windows seems to be related to path encoding (again) — what do you propose we do?

I managed to get the Windows box back and fix the issue. All the deflookups that work on macOS now work on Windows as well. There are still a few (like looking up the definition of `nat`) that don't work on either OS — we can look into this issue later.

File: INSTALL.md (hunk 6 above)
Comment (artagnon) [reply INSTALL.md]: Done, thanks.

File: src/fstar/FStar.Interactive.Ide.fs (hunk 1 above)
Comment (artagnon) [reply src/fstar/FStar.Interactive.Ide.fs]: I changed tc_one_fragment to read the filename from the env instead of command-line options (https://github.com/FStarLang/FStar/pull/1744/files#diff-0f8be5d4810e22817ca655ca5b7778feR141), because we don’t provide a command-line option in the case of Lsp. Now, the filename in the env is “<input>” in Ide because that’s what init_range used to set (https://github.com/FStarLang/FStar/pull/1744/files#diff-324a3d5d7227d9b71dbb0baba0fce7a5R1081) — however, if we change this, then the error reporting in Ide needs to be tweaked slightly to not print an elaborate message when the fname in env and repl_fname are different (https://github.com/FStarLang/FStar/pull/1744/files#diff-324a3d5d7227d9b71dbb0baba0fce7a5R63). This change to Ide.fs is hence unavoidable.
